### ISSUE TITLE
Boundary Conditions for geometries defined by discrete points

### ISSUE THREAD

**Issue by lm0bo** [lululxvi/deepxde#1963]
Hello all,

I am working with a geometry where the walls are defined by a set of discrete points rather than a continuous surface representation. I need to correctly apply Dirichlet boundary conditions using the coordinates of these wall points, but I am encountering some challenges in doing so.

Thank you in advance!

**Comment by kyouma** [comment]
Hello. I would try `PointSetBC` instead of `DirichletBC`, or `DirichletBC` with interpolation inside `func` and `on_boundary` parameters.

**Comment by lm0bo** [comment]
Thank you for your response! I have the coordinates of the points on the wall of a complex geometry, and I want to impose a Dirichlet boundary condition with a velocity of 0 at those points.

I have defined the wall as follows:
geom_wall = dde.geometry.PointCloud(wall_coords, boundary_points=wall_coords)

And the boundary condition:
bc_u_wall = dde.icbc.PointSetBC(wall_coords, np.zeros((wall_coords.shape[0], 1)), component=0)

However, the solution obtained from the network using PointSetBC for the boundary conditions is not accurate.

**Comment by kyouma** [comment]
Must your equation be solved inside the 3D geometry (volume) or on the wall (surface)? Probably you should use `dde.geometry.PointCloud(INNER_VOLUME_POINTS, boundary_points=WALL_POINTS)` and `dde.icbc.PointSetBC(WALL_POINTS,...)`?

Also, soft boundary conditions (conditions set by loss functions) cannot provide accurate solution, there will always be some residue. You can control its size using loss weights in `model.compile()`. For choosing the weights see the "_An Expert's Guide to Training Physics-informed Neural Networks_" paper. You could also try to find complex transforms to obtain hard constrains on the boundary.

**Comment by LordPontiag** [comment]
> Hello all,
> 
> I am working with a geometry where the walls are defined by a set of discrete points rather than a continuous surface representation. I need to correctly apply Dirichlet boundary conditions using the coordinates of these wall points, but I am encountering some challenges in doing so.
> 
> Thank you in advance!

Hello dear @lm0bo , as @kyouma mentioned using `PointSetBC ` is a perfect alternative for `DirichletBC`. How ever you can use following code that I wrote and used and it works correctly.

```
PINN_walls =[]
# Define boundary conditions
def boundary_wall(X, on_boundary):
    # Convert X to a NumPy array for consistency
    X = np.array(X)
    tolerance = 1e-4  # Define a small tolerance for floating-point comparison
    wall_cond = any(np.all(np.isclose(wall_coords_2D, X, atol=tolerance), axis=1))
    # Combine with the on_boundary flag
    on_wall = wall_cond & on_boundary
    if on_wall:
        PINN_walls.append(X)
        print(f"Point:",X,"is recognized on_Wall")
    return on_wall
```

and then call this function:
 // here I called it for u and v velocity components, you can extend it to your case 

```
bc_wall_u = dde.DirichletBC(geom, lambda X: 0., boundary_wall, component=0)  # u = 0 on wall
bc_wall_v = dde.DirichletBC(geom, lambda X: 0., boundary_wall, component=1)  # v = 0 on wall
```

**Comment by lm0bo** [comment]
Thank you all so much for your advice @LordPontiag @kyouma! 
Applying hard constraints on the boundary has significantly improved the results. However, I’m still unsure how to handle the case where the wall is defined by a set of points. Any suggestions on the best way to implement this?

**Comment by LordPontiag** [comment]
> Thank you all so much for your advice [@LordPontiag](https://github.com/LordPontiag) [@kyouma](https://github.com/kyouma)! Applying hard constraints on the boundary has significantly improved the results. However, I’m still unsure how to handle the case where the wall is defined by a set of points. Any suggestions on the best way to implement this?

You're most welcome!
I cannot understand your question. both ways work for implementing wall.

**Comment by lm0bo** [comment]
I apologize if my question wasn't clear. I’m not sure how to apply hard constraints when the wall is defined by a set of discrete points. 

Thank you in advance!

**Comment by kyouma** [comment]
> how to apply hard constraints when the wall is defined by a set of discrete points

Perhaps (or more possibly: for sure) my suggestion is not the best option, but you could try using a transformation $u(x) \cdot f(x) + g(x)$, where $u(x)$ is the network output; a (_perhaps "continuous" is the best option_) $f(x)=0$ for $x \in points_{BC}$; and (_perhaps not necessary "continuous"_) $g(x)=u_{BC}(x)$ for $x \in points_{BC}$. The $f(x)$ may look like $\prod_{x_i \in points_{BC}} (x - x_i)$.

I have not tested this code, but maybe it may help you. I have tried to write a backend-agnostic approach.
- `BC_COORDS` is a `NxD` Numpy array of N boundary points, each one is a vector of `D` input coordinates $(x_1, ..., x_N)$;
- `BC_VALUES` is a `NxC` Numpy array of N values of $u_{BC}(x)$ for each boundary point; `C` is the output size (e.g. for a vector-function).
See examples:
- [one](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward/poisson.dirichlet.1d.exactbc.html)
- [two](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward/diffusion.1d.exactBC.html)
and papers:
- [Exact imposition of boundary conditions with distance functions in physics-informed deep neural networks](https://doi.org/10.1016/j.cma.2021.114333)
and for periodic conditions:
- [Physics-informed neural networks with hard constraints for inverse design](https://doi.org/10.1137/21M1397908)
- [A method for representing periodic functions and enforcing exactly periodic boundary conditions with deep neural networks](https://doi.org/10.1016/j.jcp.2021.110242)
- time-periodic can be found here: [An expert's guide to training physics-informed neural networks](https://doi.org/10.48550/arXiv.2308.08468)

```
import scipy.interpolate
INTERPOLATOR = scipy.interpolate.NearestNDInterpolator(BC_COORDS, BC_VALUES)  # Or LinearNDInterpolator with fill_value!=NaN


def output_transform(coords, values):  # coords ~ MxD backend tensor; values ~ MxC backend tensor
    bc_coords = dde.backend.as_tensor(BC_COORDS, dtype=coords.dtype)  # You can also store this outside of this function for speed, but check for the dtype and device placement
    differences = dde.backend.expand_dims(coords, 1) - dde.backend.expand_dims(bc_coords, 0)  # Calculate pairwise differences of coordinates (be careful: very memory-intensive; more efficient ways must exist; you can try to find one for your backend)
    distances = dde.backend.sum(dde.backend.abs(differences), dim=-1, keepdims=False)  # You get MxN pairwise distances between input points and BC points (any type: must be 0 at x in BC_COORDS and >0 otherwise)
    
    normalized_distances = dde.backend.tanh(distances)  # Normalize to [0; 1] to avoid overflow and rounding errors during multiplication later
    # normalized_distances = dde.backend.sigmoid(distances) * 2 - 1  # Another way (I don't know which one is faster)
    remoteness_from_bc_points = dde.backend.prod(normalized_distances, dim=-1, keepdims=True)  # Mx1 array of elements: 0 when at any BC point, !=0 otherwise
    
    # OR
    
    remoteness_from_bc_points = dde.backend.min(distances, dim=-1, keepdims=True) # An alternative to "normalized_distances = ..." and "remoteness_from_bc_points = ..." (not sure which one is better)
    
    zeroed_network_output = remoteness_from_bc_points * values  # Zero the network outputs at BC points
    condition_values = INTERPOLATOR(dde.backend.to_numpy(coords))  # Any function that produces exactly BC_VALUES at corresponding BC_POINTS
    condition_values = dde.backend.as_tensor(condition_values, dtype=zeroed_network_output.dtype)
    
    return zeroed_network_output + condition_values


net.apply_output_transform(output_transform)

# Create model and train it
```
